MetaMask/metamask-extension
# Fix firsttimeloaded logic

Fixes https://github.com/MetaMask/metamask-extension/issues/18244

This PR fixes two problems:

1. A race condition in the account tracker upon service worker restart, where the `start()` method could be called before the `_loginUser()` call in the metamask-controller resolves, which means the account tracker `_updateAccounts()` could be called before there are any accounts in state, resulting in the `accounts` object in state being set to empty. This was solved by ensuring that the account-tracker state is populated during construction _if_ we are in a case where the service worker is restarting
2. The code for deciding whether "we are in a case where the service worker is restarting" was not working. We were deciding this based on the `globalThis.isFirstTimeProfileLoaded` variable. Which would be undefined in all cases, except when the browser is first started, but not when the extension is reloaded, and not upon installing the extension during a browser session that has already started. This was solved by using session storage to set whether we are in the first instance of the metamask controller being started within the current session of the extension.

To test this manually:
1. Run an mv3 build
2. Terminate the service worker a bunch of times, via chrome://inspect
3. the service worker should continuously restart successfully
